packages/remirror__extension-link/src/link-extension.ts

```diff
@@ -739,6 +803,55 @@ export class LinkExtension extends MarkExtension<LinkOptions> {
 
     return autoLinkAllowedTLDs.includes(tld);
   }
+
+  private isValidUrl(text: string, href?: string) {
+    return (
+      this.isValidTLD(href || this.buildHref(text)) && this._autoLinkRegexNonGlobal?.test(text)
+    );
+  }
+
+  private getLinkEndIndex(input: string, url: string) {
+    const domain = extractDomain(addProtocol(input, this.options.defaultProtocol));
+
+    if (domain.length === 0) {
+      return;
+    }
+
+    const path = input.slice(domain.length + input.indexOf(domain)).slice(1);
+
+    // URL API could better determine a valid path but using extractDomain appears to do the job
+    // Need to make sure that path does not include "."
+    if (!path || path.includes('.')) {
+      // Return index to remove adjacent punctuation
+      return getTrailingCharIndex({
+        adjacentPunctuations: this.options.adjacentPunctuations,
+        input,
+        url,
+      });
+    }
+
+    // A URL with a path or search query is slightly more ambiguous
+    const index = -1;
+    const balancedPathIndex = !isBalanced(path) ? getBalancedIndex(path, index) : 0;
+
+    // Balanced part of path
+    const balanced = path.slice(0, balancedPathIndex);
+    // Imbalanced part of path
+    const imbalanced = path.replace(balanced, '');
+
+    const balanceIndex = ![
+      ...imbalanced.slice((!isBalanced(imbalanced) ? getBalancedIndex(imbalanced, index) : 0) + 1),
+    ].some((char) => ![')', ']', '}'].includes(char) && !SENTENCE_PUNCTUATIONS.includes(char))
+      ? balancedPathIndex
+      : undefined;
```

Comment:
I'm struggling to understand whats happening here

Reply:
Yes sorry, let me add some comments. Also, think I can optimize this a little more.